pkg/util/json.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+func K8sJSONMerge(baseJSON, overlayJSON []byte, path Path) ([]byte, error) {
+	var base, overlay interface{}
+	err := json.Unmarshal(baseJSON, &base)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(overlayJSON, &overlay)
+	if err != nil {
+		return nil, err
+	}
+
+	dst, err := doK8sJSONMerge(base, overlay, path)
+	if err != nil {
+		return nil, err
+	}
+
+	dstJSON, err := json.Marshal(dst)
+	if err != nil {
+		return nil, err
+	}
+
+	return dstJSON, nil
+}
+
+func doK8sJSONMerge(base, overlay interface{}, path Path) (interface{}, error) {
+	var remainPath Path
+	var currentNode string
+	if len(path) == 0 {
+		currentNode = ""
+		remainPath = PathFromString("")
+	} else {
+		currentNode = path[0]
+		remainPath = path[1:]
+	}
+	switch base := base.(type) {
+	case map[string]interface{}:
+		overlay, ok := overlay.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("invalid type for current node: %s, expect object type", currentNode)
+		}
+		for k, v2 := range overlay {
+			if v1, ok := base[k]; ok {
+				merged, err := doK8sJSONMerge(v1, v2, remainPath)
+				if err != nil {
+					return nil, err
+				}
+				base[k] = merged
+			} else {
+				base[k] = v2
+			}
+		}
+	case []interface{}:
+		overlay, ok := overlay.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("invalid type for current node: %s, expect slice type", currentNode)
+		}
+		dst := base[:0]
+		dst = append(dst, base...)
+		for _, v2 := range overlay {
+			// non-leaf list, expect to match item by key.
```

Comment:
This comment is a bit confusing... for non-leaf list I think we should do a full match of all keys. For leaf list, we would select best match (the list item with most keys matching) and if nothing matches, append.